README.md
writable-stream-and-promise
===========================

[![npm](https://img.shields.io/npm/v/writable-stream-and-promise)](https://www.npmjs.com/package/writable-stream-and-promise)
[![CI](https://github.com/raymond-lam/writable-stream-and-promise/workflows/CI/badge.svg?branch=master)](https://github.com/raymond-lam/writable-stream-and-promise/actions)
[![Coverage Status](https://coveralls.io/repos/github/raymond-lam/writable-stream-and-promise/badge.svg?branch=master)](https://coveralls.io/github/raymond-lam/writable-stream-and-promise?branch=master)
[![MIT license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT)

Pipe a [stream](https://nodejs.org/api/stream.htm) into a Promise.

## Usage

### Example

```javascript
const {createReadStream} = require('fs');
const {writableStreamAndPromise} = require('writable-stream-and-promise');
const {createGunzip} = require('zlib');

const processZipFile = async path => {
  const [writable, promise] = writableStreamAndPromise();
  createReadStream(path).pipe(createGunzip()).pipe(writable);
  const contents = await promise;

  // ... process contents ...

  return processedContents;
};
```

### API

#### writableStreamAndPromise([options])

Returns a 2-tuple (Array): a [Writable](https://nodejs.org/api/stream.html#stream_writable_streams) [Buffer](https://nodejs.org/api/buffer.html) stream and a Promise. When the stream ends, the Promise resolves to a Buffer containing all the chunks written to the stream. The Promise rejects if the stream emits an error or when the stream is destroyed. Strings written to the stream will be decoded according to the given encoding or `defaultEncoding`. The stream does not destroy itself automatically after ending. The stream will emit a `'close'` event when it is destroyed.

- `options`
  - `defaultEncoding` The default encoding that is used when no encoding is specified when writing a String. The default is `'utf8'`.
  - `highWaterMark` Writes are completed immediately and synchronously, so this parameter is only really relevant if the stream is [cork](https://nodejs.org/api/stream.html#stream_writable_cork)ed. `highWaterMark` indicates the number of bytes that the stream should buffer while corked. The default is 16384 (16 kb). 
- Returns: A 2-tuple (Array of length 2), the first element a [Writable](https://nodejs.org/api/stream.html#stream_writable_streams) stream, and the second element a Promise.

#### writableObjectStreamAndPromise([options])

Returns a 2-tuple (Array): a [Writable](https://nodejs.org/api/stream.html#stream_writable_streams) [object stream](https://nodejs.org/api/stream.html#stream_object_mode) and a Promise. When the stream ends, the Promise resolves to an Array containing all the chunks written to the stream. The Promise rejects if the stream emits an error or when the stream is destroyed. The stream does not destroy itself automatically after ending. The stream will emit a `'close'` event when it is destroyed.

- options
  - `highWaterMark` Writes are completed immediately and synchronously, so this parameters is only really relevant if the stream is [cork](https://nodejs.org/api/stream.html#stream_writable_cork)ed. `highWaterMark` indicates the number of objects that the stream should buffer while corked. The default is 16.
- Returns: A 2-tuple (Array of length 2), the first element a [Writable](https://nodejs.org/api/stream.html#stream_writable_streams) stream, and the second element a Promise.

## Author

Raymond Lam (ray@lam-ray.com)

## License

MIT


CHANGELOG.md
# Changelog

## [1.0.15] - 2020-12-25

### Added

- Dependabot configuration

### Changed

- Removed all tag triggers in CI workflow (assuming that the case is covered by release publication trigger)
- Dev dependency updates

## [1.0.14] - 2020-12-25

### Fixed

- Removed duplicative tag trigger in CI workflow

### Changed

- Removed dev dependency on greenkeeper-lockfile

## [1.0.13] - 2020-12-25

### Fixed

- Updated dev dependency to address security vulnerability

### Changed

- Use GitHub Actions instead of Travis-CI for CI/CD
- Remove references to Greenkeeper.io

## [1.0.12] - 2020-07-18

### Fixed

- Updated dev dependency to address security vulnerability

## [1.0.11] - 2020-07-05

### Fixed

- Fixed bug where .end() was rejecting the Promise
- Updated dev dependencies

## [1.0.10] - 2020-04-05

### Fixed

- Updated dev dependency to address security vulnerability

## [1.0.9] - 2020-03-14

### Fixed

- Updated dev dependency to address security vulnerability

## [1.0.8] - 2020-02-09

### Changed

- Updated dev dependencies

### Fixed

- Fixed deployment

## [1.0.7] - 2019-12-29

### Changed

- Updated dev dependencies

## [1.0.6] - 2019-10-27

### Changed

- Updated dev dependencies

### Fixed

- Fixed failing lint due to updated dev dependencies

## [1.0.5] - 2019-09-02

### Changed

- Updated dev dependencies

## [1.0.4] - 2019-08-03

### Changed

- Updated dev dependencies

## [1.0.3] - 2019-05-05

### Fixed

- Fixed coveralls repo token

## [1.0.2] - 2019-05-05

### Fixed

- Removed unnecessary dev dependencies to fix failing coverage script

### Changed

- Updated dev dependencies
- minor change README.md format

### Added

- coveralls.io integration
- additional badges in README.md

## [1.0.1] - 2019-02-25

### Fixed

- Fixed `mocha` mismatch between `package.json` and `package-lock.json`
- Clarified example in `README.md`

## [1.0.0] - 2019-02-24

### Added

- Initial functionality
- Tests
- CI


test.js
/* global describe it */
/* eslint-disable no-magic-numbers */

'use strict';

const {expect} = require('chai');
const {
  writableStreamAndPromise,
  writableObjectStreamAndPromise,
} = require('./index');
const {constants: {MAX_LENGTH}} = require('buffer');

describe('writableStreamAndPromise', () => {
  it(
    'returns a writable stream and promise where the stream is immediately '
      + 'flushed when written to and the promise resolves to the concatenation '
      + 'of the Buffers written to the stream when the stream ends',
    async () => {
      const [writable, promise] = writableStreamAndPromise();

      // eslint-disable-next-line no-unused-expressions
      expect(writable.write(Buffer.from([1, 2, 3]))).to.be.true;

      // eslint-disable-next-line no-unused-expressions
      expect(writable.write(Buffer.from([6, 5, 4]))).to.be.true;
      writable.end();

      const buffer = await promise;

      // eslint-disable-next-line no-unused-expressions
      expect(Buffer.isBuffer(buffer)).to.be.true;
      expect(Array.from(buffer.values())).to.deep.equal([1, 2, 3, 6, 5, 4]);
    },
  );

  it(
    'returns a writable stream and promise where the promise resolves to the '
      + 'strings written to the stream encoded in utf-8',
    async () => {
      const [writable, promise] = writableStreamAndPromise();
      writable.write('abc');
      writable.write('x y z');
      writable.end();

      const buffer = await promise;

      // eslint-disable-next-line no-unused-expressions
      expect(Buffer.isBuffer(buffer)).to.be.true;
      expect(buffer.toString()).to.equal('abcx y z');
    },
  );

  it(
    'returns a writable stream and promise where the promise resolves to the '
      + 'strings written to the stream encoded in specified defaultEncoding',
    async () => {
      const [writable, promise] = writableStreamAndPromise({
        defaultEncoding: 'ucs-2',
      });
      writable.write('abc');
      writable.write('x y z');
      writable.end();

      const buffer = await promise;

      // eslint-disable-next-line no-unused-expressions
      expect(Buffer.isBuffer(buffer)).to.be.true;
      expect(buffer.toString('ucs-2')).to.equal('abcx y z');
    },
  );

  it(
    'returns a writable stream and promise where the promise resolves to the '
      + 'strings written to the stream encoded in the specified encoding in '
      + 'each write',
    async () => {
      const [writable, promise] = writableStreamAndPromise({
        defaultEncoding: 'ucs-2',
      });
      writable.write('abc', 'utf-8');
      writable.write('x y z');
      writable.write('bbb', 'hex');
      writable.end();

      const buffer = await promise;

      // eslint-disable-next-line no-unused-expressions
      expect(Buffer.isBuffer(buffer)).to.be.true;

      expect(Array.from(buffer.values())).to.deep.equal([
        ...Array.from(Buffer.from('abc', 'utf-8')),
        ...Array.from(Buffer.from('x y z', 'ucs-2')),
        ...Array.from(Buffer.from('bbb', 'hex')),
      ]);
    },
  );

  it(
    'returns a writable stream which buffers when corked until the '
      + 'specified high water mark is reached',
    () => new Promise(
      resolve => {
        const [writable, promise] = writableStreamAndPromise({
          highWaterMark: 36,
        });

        promise.then(buffer => {
          expect(
            Array.from(buffer.values()),
          ).to.deep.equal([
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            1, 2, 3,
            4, 5, 6,
            7, 8, 9,
          ]);
          resolve();
        });

        writable.cork();

        let totalBytes = 0;
        do {
          totalBytes += 3;
        } while (writable.write(Buffer.from([1, 2, 3])));

        expect(totalBytes).to.equal(36);

        writable.once('drain', () => {
          // eslint-disable-next-line no-unused-expressions
          expect(writable.write(Buffer.from([4, 5, 6]))).to.be.true;
          writable.end(Buffer.from([7, 8, 9]));
        });

        writable.uncork();
      },
    ),
  );

  it(
    'returns a writable stream and promise where the stream frees its '
      + 'collected values, emits a "close" event and causes the promise to '
      + 'reject when the stream is destroyed',
    async () => {
      const [writable, promise] = writableStreamAndPromise();

      const closePromise = new Promise(
        resolve => writable.once('close', resolve),
      );

      writable.write('abc');
      writable.destroy();
      await closePromise;

      // A bit of a hack: call _collect to resolve the promise to the collected
      // bytes, which would throw an error if called after destruction, since
      // the pointer to the collected bytes is set to null.
      expect(() => writable._collect(() => {})).to.throw(
        "Cannot read property 'next' of null",
      );

      let error;
      try {
        await promise;
      }
      catch (e) {
        error = e;
      }

      expect(error).to.be.an.instanceof(Error);
      expect(error.message).to.equal('Stream was destroyed.');
    },
  );

  it(
    'returns a writable stream and promise where the promise rejects with the '
      + 'error that is given to the .destroy() method of the stream',
    async () => {
      const [writable, promise] = writableStreamAndPromise();

      writable.write('abc');

      const destroyError = new Error('test message');
      writable.destroy(destroyError);

      let promiseError;
      try {
        await promise;
      }
      catch (e) {
        promiseError = e;
      }

      expect(promiseError).to.equal(destroyError);
    },
  );

  it(
    'returns a writable stream and promise where the promise rejects if all '
      + 'data written cannot be held in a single Buffer',
    async () => {
      const [writable, promise] = writableStreamAndPromise();

      writable.write(Buffer.alloc(MAX_LENGTH));
      writable.write(Buffer.alloc(1));
      writable.end();

      let error;
      try {
        await promise;
      }
      catch (e) {
        error = e;
      }

      expect(error).to.be.an.instanceof(Error);
      expect(error.message).to.equal(
        `Total length of buffered data (${MAX_LENGTH + 1} bytes) exceeds `
          + `maximum allowed length of Buffer (${MAX_LENGTH} bytes).`,
      );
    },
  );
});

describe('writableObjectStreamAndPromise', () => {
  it(
    'returns a writable stream and promise where stream is immediately flushed '
      + 'when written to and the promise resolves to an Array of the objects '
      + 'written to the stream when the stream ends',
    async () => {
      const [writable, promise] = writableObjectStreamAndPromise();

      // eslint-disable-next-line no-unused-expressions
      expect(writable.write('abc')).to.be.true;

      // eslint-disable-next-line no-unused-expressions
      expect(writable.write(512)).to.be.true;

      // eslint-disable-next-line no-unused-expressions
      expect(writable.write({foo: 'bar'})).to.be.true;

      // eslint-disable-next-line no-unused-expressions
      expect(writable.write([1, 'a', {biz: 'baz'}])).to.be.true;

      writable.end();

      const arr = await promise;

      expect(arr).to.be.an.instanceof(Array);
      expect(arr).to.deep.equal([
        'abc',
        512,
        {foo: 'bar'},
        [1, 'a', {biz: 'baz'}],
      ]);
    },
  );

  it(
    'returns a writable stream which buffers when corked until the '
      + 'specified high water mark is reached',
    () => new Promise(
      resolve => {
        const [writable, promise] = writableObjectStreamAndPromise({
          highWaterMark: 5,
        });

        promise.then(buffer => {
          expect(buffer).to.deep.equal([
            'xyz',
            'xyz',
            'xyz',
            'xyz',
            'xyz',
            {abc: 'def'},
            0,
          ]);
          resolve();
        });

        writable.cork();

        let totalObjects = 0;
        do {
          ++totalObjects;
        } while (writable.write('xyz'));

        expect(totalObjects).to.equal(5);

        writable.once('drain', () => {
          // eslint-disable-next-line no-unused-expressions
          expect(writable.write({abc: 'def'})).to.be.true;
          writable.end(0);
        });

        writable.uncork();
      },
    ),
  );

  it(
    'returns a writable stream and promise where the stream frees its '
      + 'collected values, emits a "close" event and causes the promise to '
      + 'reject when the stream is destroyed',
    async () => {
      const [writable, promise] = writableObjectStreamAndPromise();

      const closePromise = new Promise(
        resolve => writable.once('close', resolve),
      );

      writable.write('abc');
      writable.destroy();

      await closePromise;

      // A bit of a hack: call _collect to resolve the promise to the collected
      // bytes, which would throw an error if called after destruction, since
      // the pointer to the collected bytes is set to null.
      expect(() => writable._collect(() => {})).to.throw(
        "Cannot read property 'next' of null",
      );

      let error;
      try {
        await promise;
      }
      catch (e) {
        error = e;
      }

      expect(error).to.be.an.instanceof(Error);
      expect(error.message).to.equal('Stream was destroyed.');
    },
  );

  it(
    'returns a writable stream and promise where the promise rejects with the '
      + 'error that is given to the .destroy() method of the stream',
    async () => {
      const [writable, promise] = writableObjectStreamAndPromise();

      writable.write('abc');

      const destroyError = new Error('test message');
      writable.destroy(destroyError);

      let promiseError;
      try {
        await promise;
      }
      catch (e) {
        promiseError = e;
      }

      expect(promiseError).to.equal(destroyError);
    },
  );
});


index.js
'use strict';

//  writable-stream-and-promise
//  https://github.com/raymond-lam/writable-stream-and-promise
//  (c) 2019 Raymond Lam
//
//  Author: Raymond Lam (ray@lam-ray.com)
//
//  writable-stream-and-promise may be freely distributed under the MIT license.

const EventEmitter = require('events');
const {Writable} = require('stream');
const {constants: {MAX_LENGTH}} = require('buffer');

// Lock in the defaults as constants in order not to be dependent on (possibly
// changing) Node.js defaults.
const DEFAULT_ENCODING = 'utf8';
const DEFAULT_HIGH_WATER_MARK_BYTES = 16384;
const DEFAULT_HIGH_WATER_MARK_OBJECTS = 16;

// Returns a 3-tuple (Array): a base class for writableStreamAndPromise and
// writableStreamAndPromise, a reference to a linked list which will hold the
// written chunks, and an EventEmitter, which will communicate when the stream
// is destroyed or when the stream is ended and the chunks are collected.
const getBaseClassAndProtectedMembers = () => {
  // For internal communication between the stream and the promise.
  const internalEvents = new EventEmitter();

  // Store the written data as a linked list. The caller should keep a reference
  // to the head.
  let tail = {};

  // true if _final was called, and false otherwise.
  let end = false;

  return [
    class extends Writable {
      // eslint-disable-next-line class-methods-use-this
      _write(chunk, encoding, callback) {
        // Storing the written values in a linked list is faster than pushing
        // them to a variable-length array.
        tail.value = chunk;
        tail.next = {};
        tail = tail.next;

        callback();
      }

      _destroy(err, callback) { // eslint-disable-line class-methods-use-this
        // Overrided _destroy must also let go of the head of the linked list
        // for this to be effective.
        tail = null;

        if (!err && !end) {
          callback(new Error('Stream was destroyed.'));
        }
        else {
          callback(err);
        }
      }

      _final(callback) {
        let contents;
        try {
          contents = this._collect();
        }
        catch (e) {
          callback(e);
          return;
        }

        end = true;
        callback();
        internalEvents.emit('collect', contents);
      }
    },
    tail,
    internalEvents,
  ];
};

module.exports = {

  // Returns a 2-tuple (Array): a Writable Buffer stream and a Promise. When the
  // stream ends, the Promise resolves to a Buffer containing all the chunks
  // written to the stream. The Promise rejects if the stream emits an error or
  // when the stream is destroyed. Strings written to the stream will be decoded
  // according to the given encoding or defaultEncoding. The stream does not
  // destroy itself automatically after ending. The stream will emit a 'close'
  // event when it is destroyed.
  //
  // Named parameters:
  //   defaultEncoding (optional) - The default encoding that is used when no
  //                                encoding is specified when writing a String.
  //                                The default is 'utf8'.
  //   highWaterMark (optional) -   Writes are completed immediately and
  //                                synchronously, so this parameter is only
  //                                really relevant if the stream is corked.
  //                                highWaterMark indicates the number of bytes
  //                                that the stream should buffer while corked.
  //                                The default is 16384 (16 kb).
  writableStreamAndPromise({
    highWaterMark = DEFAULT_HIGH_WATER_MARK_BYTES,
    defaultEncoding = DEFAULT_ENCODING,
  } = {}) {
    let [BaseClass, head, internalEvents] = getBaseClassAndProtectedMembers();
    let totalLength = 0;

    const writable = new class extends BaseClass {
      _write(chunk, encoding, callback) {
        totalLength += chunk.length;
        super._write(chunk, encoding, callback);
      }

      _destroy(err, callback) {
        head = null;
        super._destroy(err, callback);
      }

      _collect() { // eslint-disable-line class-methods-use-this
        if (totalLength > MAX_LENGTH) {
          throw new Error(
            `Total length of buffered data (${totalLength} bytes) exceeds `
              + `maximum allowed length of Buffer (${MAX_LENGTH} bytes).`,
          );
        }

        // Buffer.allocUnsafe is faster than Buffer.alloc, but it is important
        // to ensure that it is completely filled before returning, in order
        // not to expose uninitialized (and potentially sensitive) parts of
        // memory.
        const buffer = Buffer.allocUnsafe(totalLength);
        let offset = 0;
        for (; head.next; head = head.next) {

          // Traditional for loop is the fastest way to iterate on the values
          // of a Buffer.
          for (let i = 0; i < head.value.length; ++i) {
            buffer[offset] = head.value[i];
            ++offset;
          }
        }

        // Ensure that the chunks fill the allocated buffer exactly.
        /* istanbul ignore next */if (offset !== totalLength) {
          throw new Error(
            'length property of chunks do not sum to the length of the Buffer '
              + 'allocated for them.',
          );
        }

        return buffer;
      }
    }({highWaterMark, defaultEncoding});

    return [writable, new Promise(
      (resolve, reject) => {
        writable.once('error', reject);
        internalEvents.once('collect', resolve);
      },
    )];
  },

  // Returns a 2-tuple (Array): a Writable object stream and a Promise. When the
  // stream ends, the Promise resolves to an Array containing all the chunks
  // written to the stream. The Promise rejects if the stream emits an error or
  // when the stream is destroyed. The stream does not destroy itself
  // automatically after ending. The stream will emit a 'close' event when it is
  // destroyed.
  //
  // Named parameters:
  //   highWaterMark (optional) -   Writes are completed immediately and
  //                                synchronously, so this parameters is only
  //                                really relevant if the stream is corked.
  //                                highWaterMark indicates the number of
  //                                objects that the stream should buffer while
  //                                corked. The default is 16.
  writableObjectStreamAndPromise({
    highWaterMark = DEFAULT_HIGH_WATER_MARK_OBJECTS,
  } = {}) {
    let [BaseClass, head, internalEvents] = getBaseClassAndProtectedMembers();
    let totalLength = 0;

    const writable = new class extends BaseClass {
      _write(chunk, encoding, callback) {
        ++totalLength;
        super._write(chunk, encoding, callback);
      }

      _destroy(err, callback) {
        head = null;
        super._destroy(err, callback);
      }

      _collect() { // eslint-disable-line class-methods-use-this

        // Writing to a pre-allocated array is faster than pushing onto a
        // dynamically sized array.
        const arr = new Array(totalLength);

        let offset = 0;
        for (; head.next; head = head.next) {
          arr[offset] = head.value;
          ++offset;
        }

        return arr;
      }
    }({objectMode: true, highWaterMark});

    return [writable, new Promise(
      (resolve, reject) => {
        writable.once('error', reject);
        internalEvents.once('collect', resolve);
      },
    )];
  },
};
